### ISSUE TITLE
Better exception message when configuring/asserting a non-fake object

### ISSUE THREAD

**Issue by adamralph** [FakeItEasy/FakeItEasy#682]
*Extracted from #681*

Currently:

```
The current proxy generator can not intercept the specified method for the following reason:
    - Sealed methods can not be intercepted.
```

better would be something like

```
The current proxy generator can not intercept the specified method for the following reason:
    - The object is not a fake.
```

**Comment by adamralph** [comment]
Is the exception message I'm suggesting best/possible?

**Comment by blairconrad** [comment]
I'd thought that we might want to shut down the operation earlier in the chain. If we know that the instance referred to in `A.CallTo[Set]` is not a Fake, we could abort at that time, long before even getting to a proxy. Really I'm not sure how we got down to the DynamicProxyGenerator anyhow - if the object isn't a Fake, how did we really know which generator (Dynamic or Delegate) to use?

If we can get the info, what about

> Cannot configure or assert calls on <object name>. It is not a Fake.

**Comment by adamralph** [comment]
Sounds good to me! Yes, figuring this out earlier sounds doable and sensible.

**Comment by blairconrad** [comment]
I'd start looking in `FakeConfigurationManager.AssertThatMemberCanBeIntercepted`.

**Comment by blairconrad** [comment]
This issue appears to be slightly different than I thought. I had thought that we'd get a bad exception message any time we try to configure a non-fake, but it turns out that

```c#
[Scenario]
public static void CallToNotAFake(
    BaseClass notAFake,
    Exception exception)
{
    "Given an object that is not a fake"
        .x(() => notAFake = new BaseClass());

    "When I start to configure a method on the fake"
        .x(() => exception = Record.Exception(() => A.CallTo(() => notAFake.DoSomething())));

    "Then it throws an invalid operation exception"
        .x(() => exception.Should().BeNull());
}
```

yields:

```
Expected object to be <null>, but found System.ArgumentException with message "The specified object is not recognized as a fake object."
```

Which is pretty good. The problem may be specific to the extension method approach that @TimLovellSmitth outlined in #681.

**Comment by blairconrad** [comment]
Okay. It's convoluted, but I can reproduce the behaviour with an actual sealed method that's exercised via an extension method.

```c#
public class BaseClass
{
    public bool WasCalled { get; private set; }

    public virtual void DoSomething()
    {
        this.WasCalled = true;
    }

    public virtual int ReturnSomething()
    {
        this.WasCalled = true;
        return 10;
    }
}

public class DerivedClass: BaseClass
{
    public sealed override void DoSomething()
    {
    }
}

public static class BaseClassExtensions
{
    public static void DoSomething(this ConfigurationSpecs.BaseClass subject, string dummy)
    {
        subject.DoSomething();
    }
}

[Scenario]
public static void CheckMethodCalledViaExtensionMethodOnNotAFake(
    BaseClass notAFake,
    Exception exception)
{
    "Given an object that is not a fake"
        .x(() => notAFake = new DerivedClass());

    "And an extension method with the same name as a method on the object"
        .See(nameof(BaseClassExtensions.DoSomething));

    "When I check to see if a method was called on the object"
        .x(() => exception = Record.Exception(() => A.CallTo(() => notAFake.DoSomething()).MustHaveHappened()));

    "Then it throws an invalid operation exception"
        .x(() => exception.Should().BeNull());
}
```

produces

```
Expected object to be <null>, but found FakeItEasy.Configuration.FakeConfigurationException with message "

  The current proxy generator can not intercept the specified method for the following reason:
    - Sealed methods can not be intercepted.
```

which comes from [CastleDynamicProxyInterceptionValidator](https://github.com/FakeItEasy/FakeItEasy/blob/492fab97d84d243415d6ab6f9daad7a7896fc267/src/FakeItEasy/Creation/CastleDynamicProxy/CastleDynamicProxyInterceptionValidator.cs#L37). You can see similar checks nearby for non-virtual methods, static methods, and so on.

**Comment by blairconrad** [comment]
Okay. The problem appears (to me) to be that we investigate the methods' characteristics before checking to see if the fake is a fake. Then we check to see if it's a fake afterward. I think the messages get clearer if we check for fakeness first. I'll probably amend `FakeConfigurationManager.AssertThatMemberCanBeIntercepted`.

**Comment by blairconrad** [comment]
I'm a moron. You don't need the extension method. You just need a non-virtual method. :blush:

**Comment by blairconrad** [comment]
Thanks, folks!

**Comment by thomaslevesque** [comment]
Released in [2.2.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/2.2.0), thanks @blairconrad!

**Comment by blairconrad** [comment]
Oh, I didn't know I got thanks. Thank you, sir!
